# 47장 에러 처리

### 에러 처리의 필요성

에러가 발생하지 않는 코드를 작성하는 것은 불가능하다. 발생한 에러에 대처하지 않고 방치하면 프로그램은 강제 종료된다.

```jsx
try{
	foo()
}catch(error){
	console.error('[에러 발생]', error);
}
```

### try…catch…finally 문

기본적으로 에러처리하는 방법은 두 가지가 있다. 단축 평가 또는 옵셔널 체이닝 연산자를 통해 확인해서 처리를 하는 방법과 에러 처리코드를 미리 등록해두고 에러가 발생하면 에러 처리 코드로 점프하도록 하는 방법이 있다.

```jsx
try{
	// 실행할 코드
	foo();
} catch(err) {
	console.error(err);
} finally {
	console.log('finally');
}
```

### Error 객체

```jsx
const error = new Error('invalid');
```

자바스크립트는 Error 생성자 함수를 포함해 7가지의 에러 객체를 생성할 수 있는 Error 생성자 함수를 제공한다.

Error : 일반적 에러 객체

SyntaxError : 자바스크립트 문법에 맞지 않는 문을 해석할 때 발생하는 에러 객체

ReferenceError : 참조할 수 없는 식별자를 참조했을 때 발생하는 에러 객체

TypeError : 피연산자 또는 인수의 데이터 타입이 유효하지 않을 때 발생하는 에러  객체

RangeError : 숫자값의 허용 범위를 벗어났을 때 발생하는 에러 객체

URIError : encodeURI 또는 decodeURI 함수에 부적절한 인수를 전달했을 때 발생하는 에러 객체

EvalError : eval 함수에서 발생하는 에러 객체

### throw 문

에러를 발생시키려면 try 코드 블록에서 throw 문으로 에러객체를 던져야한다.

```jsx
throw 표현식;
```

```jsx
const repeat = (n, f) => {
	if( typeof f !== 'function') throw new TypeError('f must be a function');

	for( var i = 0; i<n; i++){
		f(i);
	}
};
try{
	repeat(2, 1);
}catch(err){
	console.error(err);
}
```

### 에러의 전파

에러는 호출자 방향으로 전파된다.

그러나 비동기 함수인 setTimeout이나 프로미스 후속 처리 메서드의 콜백 함수는 호출자가 없다는 것이다.